{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nobackup1b/users/thiberio/eggNOG\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import itertools\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import igraph as ig\n",
    "from IPython.display import HTML\n",
    "import multiprocessing\n",
    "import pickle as pkl\n",
    "import random\n",
    "from collections import Counter\n",
    "import os\n",
    "import subprocess\n",
    "from scipy.stats import mannwhitneyu\n",
    "import ete3\n",
    "from copy import deepcopy\n",
    "from math import ceil\n",
    "\n",
    "ncbi = ete3.NCBITaxa()\n",
    "%run assess_connections-functions.ipynb\n",
    "\n",
    "%cd ~/work/eggNOG/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_genomes = pd.read_csv('../kelsey/genomes.tab',\n",
    "                              sep='\\t',\n",
    "                              index_col=0)\n",
    "\n",
    "lineages = pd.DataFrame()\n",
    "for taxid in sampled_genomes.species_taxid.unique():\n",
    "    if pd.isna(taxid):\n",
    "        continue\n",
    "    lineages = lineages.append({tax_rank: tmp_taxid \n",
    "                                 for tmp_taxid, tax_rank in ncbi.get_rank(ncbi.get_lineage(taxid)).items()},\n",
    "                                ignore_index=True)\n",
    "lineages = lineages.reindex(columns=['class', 'family',  'genus', 'phylum',\n",
    "                                     'order', 'species', 'superkingdom']).copy()\n",
    "lineages = lineages.query('superkingdom == 2').copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sampled_phyla = [phylum for phylum in lineages.phylum.unique().astype(int) if phylum > 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "working_groups  = pd.read_parquet('working_eggNOG_groups.parquet', engine='fastparquet')\n",
    "working_trees   = pd.read_parquet('working_eggNOG_trees.parquet' , engine='fastparquet')\n",
    "eggNOG_taxonomy = pd.read_parquet('eggNOG_taxonomy.parquet'      , engine='fastparquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('all_results.pkl', 'rb') as _:\n",
    "    phylum_connections = pkl.load(_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "regular_connections     = []\n",
    "significant_connections = []\n",
    "\n",
    "for group_id, group_clusters in phylum_connections:\n",
    "    if not group_clusters:\n",
    "        continue\n",
    "    \n",
    "    for cluster_num, cluster_data in group_clusters.items():\n",
    "\n",
    "        for ref_phylum, phyla_dists in cluster_data.items():\n",
    "            \n",
    "            if not phyla_dists['df'].shape[0]:\n",
    "                continue\n",
    "\n",
    "            closest_phylum = phyla_dists['df'].iloc[0, 0]\n",
    "            if not closest_phylum in sampled_phyla:\n",
    "                continue\n",
    "\n",
    "            if phyla_dists['significant']:\n",
    "                significant_connections.append( (ref_phylum,\n",
    "                                                 phyla_dists['df'].iloc[0, 0],\n",
    "                                                 phyla_dists['df'].iloc[0, 1], \n",
    "                                                 '%s#%i' % (group_id, cluster_num)) )\n",
    "\n",
    "                for index, tmp_series in phyla_dists['df'].iloc[1:, :].iterrows():\n",
    "                    if tmp_series.phylum in sampled_phyla:\n",
    "                        regular_connections.append( (ref_phylum,\n",
    "                                                         tmp_series.phylum,\n",
    "                                                         tmp_series['median'], \n",
    "                                                         '%s#%i' % (group_id, cluster_num)) )\n",
    "\n",
    "            else:\n",
    "                for index, tmp_series in phyla_dists['df'].iterrows():\n",
    "                    if tmp_series.phylum in sampled_phyla:\n",
    "                        regular_connections.append( (ref_phylum,\n",
    "                                                         tmp_series.phylum,\n",
    "                                                         tmp_series['median'], \n",
    "                                                         '%s#%i' % (group_id, cluster_num)) )\n",
    "\n",
    "significant_df = pd.DataFrame(data=significant_connections, columns=['source', 'target', 'distance', 'group'])\n",
    "regular_df     = pd.DataFrame(data=regular_connections,     columns=['source', 'target', 'distance', 'group'])\n",
    "\n",
    "significant_df.dropna(how='any', inplace=True)\n",
    "regular_df.dropna(    how='any', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggNOG_taxonomy.loc[(eggNOG_taxonomy['class']!=28211) & (eggNOG_taxonomy['phylum']==1224), 'phylum'] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>phylum</th>\n",
       "      <th>order</th>\n",
       "      <th>species</th>\n",
       "      <th>superkingdom</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1089551</th>\n",
       "      <td>28211.0</td>\n",
       "      <td>2066434.0</td>\n",
       "      <td>489140.0</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>204441.0</td>\n",
       "      <td>404900.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1089552</th>\n",
       "      <td>28211.0</td>\n",
       "      <td>41295.0</td>\n",
       "      <td>85274.0</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>204441.0</td>\n",
       "      <td>1087.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622637</th>\n",
       "      <td>28211.0</td>\n",
       "      <td>31993.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>356.0</td>\n",
       "      <td>622637.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536019</th>\n",
       "      <td>28211.0</td>\n",
       "      <td>69277.0</td>\n",
       "      <td>68287.0</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>356.0</td>\n",
       "      <td>593909.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46429</th>\n",
       "      <td>28211.0</td>\n",
       "      <td>41297.0</td>\n",
       "      <td>165695.0</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>204457.0</td>\n",
       "      <td>46429.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236908</th>\n",
       "      <td>28211.0</td>\n",
       "      <td>942.0</td>\n",
       "      <td>953.0</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>766.0</td>\n",
       "      <td>77038.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114615</th>\n",
       "      <td>28211.0</td>\n",
       "      <td>41294.0</td>\n",
       "      <td>374.0</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>356.0</td>\n",
       "      <td>114615.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434131</th>\n",
       "      <td>28211.0</td>\n",
       "      <td>942.0</td>\n",
       "      <td>33993.0</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>766.0</td>\n",
       "      <td>950.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1220582</th>\n",
       "      <td>28211.0</td>\n",
       "      <td>82115.0</td>\n",
       "      <td>357.0</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>356.0</td>\n",
       "      <td>28099.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1432055</th>\n",
       "      <td>28211.0</td>\n",
       "      <td>433.0</td>\n",
       "      <td>1434011.0</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>204441.0</td>\n",
       "      <td>215221.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>580 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           class     family      genus  phylum     order   species  \\\n",
       "index                                                                \n",
       "1089551  28211.0  2066434.0   489140.0  1224.0  204441.0  404900.0   \n",
       "1089552  28211.0    41295.0    85274.0  1224.0  204441.0    1087.0   \n",
       "622637   28211.0    31993.0      133.0  1224.0     356.0  622637.0   \n",
       "536019   28211.0    69277.0    68287.0  1224.0     356.0  593909.0   \n",
       "46429    28211.0    41297.0   165695.0  1224.0  204457.0   46429.0   \n",
       "...          ...        ...        ...     ...       ...       ...   \n",
       "1236908  28211.0      942.0      953.0  1224.0     766.0   77038.0   \n",
       "114615   28211.0    41294.0      374.0  1224.0     356.0  114615.0   \n",
       "434131   28211.0      942.0    33993.0  1224.0     766.0     950.0   \n",
       "1220582  28211.0    82115.0      357.0  1224.0     356.0   28099.0   \n",
       "1432055  28211.0      433.0  1434011.0  1224.0  204441.0  215221.0   \n",
       "\n",
       "         superkingdom  \n",
       "index                  \n",
       "1089551           2.0  \n",
       "1089552           2.0  \n",
       "622637            2.0  \n",
       "536019            2.0  \n",
       "46429             2.0  \n",
       "...               ...  \n",
       "1236908           2.0  \n",
       "114615            2.0  \n",
       "434131            2.0  \n",
       "1220582           2.0  \n",
       "1432055           2.0  \n",
       "\n",
       "[580 rows x 7 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eggNOG_taxonomy[(eggNOG_taxonomy['class']==28211) & (eggNOG_taxonomy['phylum']==1224)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# fastTree\n",
    "#\n",
    "\n",
    "def assess_transfers_between_phyla(related_phyla):\n",
    "\n",
    "    candidates    = set(significant_df.query('source==\"%i\" and target==\"%i\"' % related_phyla).group.values).union(\n",
    "                        significant_df.query('target==\"%i\" and source==\"%i\"' % related_phyla).group.values\n",
    "                    )\n",
    "    descriptions  = {f'from {related_phyla[0]} to {related_phyla[1]}': set(),\n",
    "                     f'from {related_phyla[1]} to {related_phyla[0]}': set(),\n",
    "                      'sisters':                                       set(),\n",
    "                      'no clear relationship between phyla':           set()}\n",
    "\n",
    "    for candidate in candidates:\n",
    "\n",
    "        group_id, cluster_num = candidate.split('#')\n",
    "\n",
    "        with cd('candidates/trees/'):\n",
    "            #\n",
    "            # test if there is rooted version of the tree (mad output) \n",
    "            if not os.path.isfile('%s-cluster%s.fastTree.rooted' % (group_id, cluster_num)):\n",
    "                continue\n",
    "\n",
    "            #\n",
    "            # we need the original tree cause mad removes support values, so we just transfer root positions\n",
    "            try:\n",
    "                tmp_tree   = match_rooting(ete3.Tree('%s-cluster%s.fastTree.rooted' % (group_id, cluster_num)),\n",
    "                                           ete3.Tree('%s-cluster%s.fastTree'        % (group_id, cluster_num)))\n",
    "            except ete3.parser.newick.NewickError:\n",
    "                continue\n",
    "\n",
    "        #\n",
    "        # extract taxIDs to subsample taxonomy tables\n",
    "        taxids = {}\n",
    "        for leaf in tmp_tree.get_leaf_names():\n",
    "            #\n",
    "            # leaf names are composed by <taxid>.<locus_tag>\n",
    "            #   ps: locus_tag may also have \".\" within it\n",
    "            tmp_taxid = int(leaf.split('.')[0])\n",
    "\n",
    "            if tmp_taxid not in taxids:\n",
    "                taxids[tmp_taxid] = []\n",
    "            taxids[tmp_taxid].append(leaf)\n",
    "\n",
    "        #\n",
    "        # select taxIDs from each assessed phylum...\n",
    "        phylum1_taxonomy = eggNOG_taxonomy.loc[taxids].query('phylum==@related_phyla[0]')\n",
    "        phylum2_taxonomy = eggNOG_taxonomy.loc[taxids].query('phylum==@related_phyla[1]')\n",
    "\n",
    "        #\n",
    "        # ... and their respect leaves\n",
    "        phylum1_leaves = set()\n",
    "        for taxid in phylum1_taxonomy.index:\n",
    "            phylum1_leaves.update(taxids[taxid])\n",
    "\n",
    "        phylum2_leaves = set()\n",
    "        for taxid in phylum2_taxonomy.index:\n",
    "            phylum2_leaves.update(taxids[taxid])\n",
    "        \n",
    "        all_leaves = phylum1_leaves.union(phylum2_leaves)\n",
    "\n",
    "        #\n",
    "        # as tree traversing through ete3 isn't very efficient, and doesn't scalate very well,\n",
    "        #   we create an iGraph dag for more effient traversing\n",
    "        dag  = tree_to_dag(tmp_tree)          # we need a directed version...\n",
    "        udag = dag.as_undirected(mode='each') # ... and an undirected one for different processes\n",
    "\n",
    "        #\n",
    "        # placeholders where we will add monophyletic clades for each phylum\n",
    "        phylum1_clades = set() \n",
    "        phylum2_clades = set()\n",
    "\n",
    "        ignored_nodes  = [] # descendants of monophyletic nodes should be ignored once their\n",
    "                            #   ancestors have been stored.\n",
    "\n",
    "        #\n",
    "        # traverse through internal nodes\n",
    "        for node in dag.vs:\n",
    "            if node.index in ignored_nodes:\n",
    "                continue\n",
    "\n",
    "            node_leaves = get_leaf_names(node)\n",
    "\n",
    "            if all_leaves.isdisjoint(node_leaves):\n",
    "                ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "            #\n",
    "            # if there aren't leaves from other phyla within this node it is monophyletic\n",
    "            if phylum1_leaves.intersection(node_leaves) and phylum1_leaves.issuperset(node_leaves):\n",
    "                phylum1_clades.add(node.index)\n",
    "                ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "            elif phylum2_leaves.intersection(node_leaves) and phylum2_leaves.issuperset(node_leaves):\n",
    "                phylum2_clades.add(node.index)\n",
    "                ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "        #\n",
    "        # now we add some flexibility to the monophyly of nodes for three reasons:\n",
    "        #   1) donor nodes within gene trees will never be monophyletic since the recipient\n",
    "        #      must be nested within it.\n",
    "        #   2) if there are other transfers from the donor and/or recipient phyla to a 3rd one\n",
    "        #      we still want to capture it.\n",
    "        #   3) good ol' phylogenetic uncertainty\n",
    "        phylum1_clades = merge_polyphyletic_clades(phylum1_clades, udag)\n",
    "        phylum2_clades = merge_polyphyletic_clades(phylum2_clades, udag)\n",
    "\n",
    "        #\n",
    "        # flag if we can identify relations between phyla:\n",
    "        #   1 nested within 2\n",
    "        #   2 nested within 1\n",
    "        #   1 and 2 are sisters\n",
    "        #\n",
    "        # if no relation is identified, leave as false\n",
    "        phyla_relationship_flag = 0\n",
    "\n",
    "        for clade1, clade2 in itertools.product(phylum1_clades, phylum2_clades):\n",
    "\n",
    "            #\n",
    "            # capture ancestors of clade1 by querying nodes between itself and the root node\n",
    "            clade1_ancestors = udag.vs[clade1].get_shortest_paths(udag.vs[0])[0][1:]\n",
    "            #\n",
    "            # if clade2 within clade1 ancestors it means that clade1 is nested within clade2\n",
    "            #   evidence of transfer from clade2 -> clade1\n",
    "            if clade2 in clade1_ancestors:\n",
    "#                 support = dag.vs[clade2].in_edges()[0]['support']\n",
    "#                 print('%i is nested within %i             (%s)' % (related_phyla[0], related_phyla[1], candidate), support)\n",
    "                recipient_taxon         = fine_tune_recipient(dag.vs[clade1], \n",
    "                                                              related_phyla[0])\n",
    "                donor_taxon             = fine_tune_recipient(dag.vs[clade2], \n",
    "                                                              related_phyla[1])\n",
    "                phyla_relationship_flag = 1\n",
    "\n",
    "                if (recipient_taxon, donor_taxon) != related_phyla:\n",
    "                    descriptions[f'from {related_phyla[1]} to {related_phyla[0]}'].add('%s from %i to %i' % \\\n",
    "                                                                                       (candidate, donor_taxon, \n",
    "                                                                                        recipient_taxon))\n",
    "                else:\n",
    "                    descriptions[f'from {related_phyla[1]} to {related_phyla[0]}'].add(candidate)\n",
    "                continue\n",
    "\n",
    "            #\n",
    "            # if clade1 within clade2 ancestors it means that clade2 is nested within clade1\n",
    "            #   evidence of transfer from clade1 -> clade2\n",
    "            clade2_ancestors = udag.vs[clade2].get_shortest_paths(udag.vs[0])[0][1:]\n",
    "            if clade1 in clade2_ancestors:\n",
    "#                 support = dag.vs[clade1].in_edges()[0]['support']\n",
    "#                 print('%i is nested within %i             (%s)' % (related_phyla[1], related_phyla[0], candidate), support)\n",
    "                recipient_taxon         = fine_tune_recipient(dag.vs[clade2], \n",
    "                                                              related_phyla[1])\n",
    "                donor_taxon             = fine_tune_recipient(dag.vs[clade1], \n",
    "                                                              related_phyla[0])\n",
    "                phyla_relationship_flag = 1\n",
    "\n",
    "                if (donor_taxon, recipient_taxon) !=  related_phyla:\n",
    "                    descriptions[f'from {related_phyla[0]} to {related_phyla[1]}'].add('%s from %i to %i' % \\\n",
    "                                                                                       (candidate, donor_taxon, \n",
    "                                                                                        recipient_taxon))\n",
    "                else:\n",
    "                    descriptions[f'from {related_phyla[0]} to {related_phyla[1]}'].add(candidate)\n",
    "                continue\n",
    "\n",
    "            #\n",
    "            # if clade1 and clade2 are have the same parent noe it means they are sisters\n",
    "            #   evidence of hgt, but information about directionality\n",
    "            if clade1_ancestors[0] == clade2_ancestors[0]:\n",
    "#                 if dag.vs[clade1].predecessors()[0].index:\n",
    "#                     support = dag.vs[clade1].predecessors()[0].in_edges()[0]['support']\n",
    "#                     print('%i and %i are sisters              (%s)' % (related_phyla[0], related_phyla[1], candidate), support)\n",
    "                descriptions['sisters'].add(candidate)\n",
    "#                 else:\n",
    "#                     print('%i and %i are sisters              (%s)' % (related_phyla[0], related_phyla[1], candidate), 'root')\n",
    "                phyla_relationship_flag = 1\n",
    "                continue\n",
    "\n",
    "    #     with cd('candidates/'):\n",
    "    #         with open('%s-cluster%s.figTree' % (group_id, cluster_num), 'w') as out:\n",
    "    #             out.write(visualize_reconstruct_candidate(tmp_tree))\n",
    "\n",
    "        #\n",
    "        # if the flag still is False, we couldn't identify an relationship between phyla\n",
    "        if not phyla_relationship_flag:\n",
    "            descriptions['no clear relationship between phyla'].add(candidate)\n",
    "\n",
    "    return(descriptions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'from 1224 to 1117': {'2Z7ZS#0 from 28211 to 1117',\n",
       "  '2Z9FC#0 from 28211 to 1117',\n",
       "  '2ZPQC#0 from 28211 to 1117',\n",
       "  '30CE8#0 from 28211 to 1117',\n",
       "  '31ARV#1 from 356 to 1117',\n",
       "  '32BAX#1 from 28211 to 1117',\n",
       "  '32BAX#1 from 356 to 1117',\n",
       "  '32TNJ#0 from 28211 to 1117',\n",
       "  '32UZ7#0 from 28211 to 1117',\n",
       "  '32WB4#0 from 28211 to 1117',\n",
       "  '32Z2Q#2 from 356 to 1117',\n",
       "  '330QS#1 from 28211 to 1117',\n",
       "  'COG0001#0 from 28211 to 1117',\n",
       "  'COG0002#0 from 28211 to 1117',\n",
       "  'COG0006#0 from 28211 to 1117',\n",
       "  'COG0031#0 from 28211 to 1416614',\n",
       "  'COG0033#0 from 28211 to 1117',\n",
       "  'COG0033#0 from 356 to 1117',\n",
       "  'COG0068#3 from 28211 to 1117',\n",
       "  'COG0119#7 from 28211 to 1117',\n",
       "  'COG0129#8 from 28211 to 1117',\n",
       "  'COG0141#3 from 28211 to 1117',\n",
       "  'COG0145#0 from 28211 to 1117',\n",
       "  'COG0146#2 from 28211 to 1117',\n",
       "  'COG0153#0 from 28211 to 1117',\n",
       "  'COG0155#5 from 28211 to 1117',\n",
       "  'COG0161#1 from 28211 to 1117',\n",
       "  'COG0166#1 from 28211 to 1117',\n",
       "  'COG0182#1 from 28211 to 1117',\n",
       "  'COG0191#3 from 28211 to 1117',\n",
       "  'COG0220#1 from 28211 to 1117',\n",
       "  'COG0226#3 from 28211 to 1117',\n",
       "  'COG0229#0 from 28211 to 1117',\n",
       "  'COG0240#1 from 28211 to 1117',\n",
       "  'COG0242#2 from 28211 to 1117',\n",
       "  'COG0260#0 from 28211 to 1117',\n",
       "  'COG0288#0 from 28211 to 1117',\n",
       "  'COG0298#0 from 356 to 1117',\n",
       "  'COG0302#4 from 28211 to 1117',\n",
       "  'COG0308#0 from 28211 to 33071',\n",
       "  'COG0315#3 from 28211 to 1117',\n",
       "  'COG0350#2 from 28211 to 1117',\n",
       "  'COG0354#1 from 28211 to 1117',\n",
       "  'COG0361#2 from 28211 to 1117',\n",
       "  'COG0380#0 from 28211 to 1117',\n",
       "  'COG0398#0 from 28211 to 1117',\n",
       "  'COG0401#0 from 28211 to 1117',\n",
       "  'COG0401#1 from 28211 to 1117',\n",
       "  'COG0402#4 from 28211 to 1117',\n",
       "  'COG0409#2 from 28211 to 1117',\n",
       "  'COG0418#0 from 28211 to 1117',\n",
       "  'COG0428#2 from 28211 to 1117',\n",
       "  'COG0431#3 from 28211 to 1117',\n",
       "  'COG0450#0 from 28211 to 1117',\n",
       "  'COG0465#13 from 28211 to 1117',\n",
       "  'COG0483#2 from 28211 to 1117',\n",
       "  'COG0511#3 from 28211 to 1117',\n",
       "  'COG0514#2 from 28211 to 1117',\n",
       "  'COG0521#2 from 28211 to 1117',\n",
       "  'COG0628#9 from 28211 to 1117',\n",
       "  'COG0633#3 from 28211 to 1117',\n",
       "  'COG0668#5 from 28211 to 1117',\n",
       "  'COG0680#5 from 28211 to 1117',\n",
       "  'COG0684#1 from 28211 to 1117',\n",
       "  'COG0695#2 from 28211 to 1117',\n",
       "  'COG0695#4 from 28211 to 1117',\n",
       "  'COG0709#3 from 28211 to 1117',\n",
       "  'COG0725#0 from 28211 to 1117',\n",
       "  'COG0727#0 from 28211 to 1117',\n",
       "  'COG0748#5 from 28211 to 1117',\n",
       "  'COG0754#3 from 28211 to 1117',\n",
       "  'COG0783#0 from 28211 to 1117',\n",
       "  'COG0863#4 from 28211 to 1117',\n",
       "  'COG0863#4 from 28211 to 1416614',\n",
       "  'COG1045#5 from 28211 to 1117',\n",
       "  'COG1054#0 from 28211 to 1117',\n",
       "  'COG1056#1 from 28211 to 1117',\n",
       "  'COG1062#3 from 28211 to 1117',\n",
       "  'COG1083#2 from 28211 to 1117',\n",
       "  'COG1112#7 from 28211 to 1117',\n",
       "  'COG1115#0 from 28211 to 1117',\n",
       "  'COG1156#2 from 31989 to 1117',\n",
       "  'COG1177#5 from 28211 to 1117',\n",
       "  'COG1178#0 from 28211 to 1117',\n",
       "  'COG1230#0 from 28211 to 1117',\n",
       "  'COG1234#0 from 28211 to 1117',\n",
       "  'COG1260#1 from 28211 to 1117',\n",
       "  'COG1262#6 from 28211 to 1117',\n",
       "  'COG1271#0 from 28211 to 1117',\n",
       "  'COG1285#2 from 28211 to 1117',\n",
       "  'COG1305#3 from 28211 to 1117',\n",
       "  'COG1317#2 from 28211 to 1117',\n",
       "  'COG1322#6 from 204799 to 1117',\n",
       "  'COG1348#3 from 28211 to 1117',\n",
       "  'COG1393#0 from 28211 to 1117',\n",
       "  'COG1402#3 from 28211 to 1117',\n",
       "  'COG1433#1 from 28211 to 1117',\n",
       "  'COG1433#3 from 28211 to 1117',\n",
       "  'COG1505#0 from 28211 to 1117',\n",
       "  'COG1524#5 from 28211 to 1117',\n",
       "  'COG1544#0 from 28211 to 1117',\n",
       "  'COG1598#1 from 28211 to 1117',\n",
       "  'COG1598#3 from 28211 to 1117',\n",
       "  'COG1682#0 from 28211 to 1117',\n",
       "  'COG1682#10 from 28211 to 1117',\n",
       "  'COG1723#3 from 28211 to 1117',\n",
       "  'COG1742#2 from 28211 to 1117',\n",
       "  'COG1742#2 from 28211 to 33072',\n",
       "  'COG1791#2 from 28211 to 1117',\n",
       "  'COG1805#1 from 28211 to 1117',\n",
       "  'COG1814#5 from 28211 to 1117',\n",
       "  'COG1819#1 from 28211 to 1117',\n",
       "  'COG1881#1 from 28211 to 1117',\n",
       "  'COG1916#5 from 28211 to 1117',\n",
       "  'COG1932#1 from 28211 to 1117',\n",
       "  'COG1938#0 from 28211 to 1117',\n",
       "  'COG1957#3 from 28211 to 1117',\n",
       "  'COG2020#5 from 28211 to 1117',\n",
       "  'COG2020#6 from 28211 to 1117',\n",
       "  'COG2032#3 from 28211 to 1117',\n",
       "  'COG2040#1 from 28211 to 1117',\n",
       "  'COG2044#3 from 28211 to 1117',\n",
       "  'COG2082#2 from 28211 to 1117',\n",
       "  'COG2099#1 from 28211 to 1117',\n",
       "  'COG2113#0 from 28211 to 1117',\n",
       "  'COG2113#2 from 28211 to 1117',\n",
       "  'COG2130#4 from 28211 to 1117',\n",
       "  'COG2214#2 from 28211 to 1117',\n",
       "  'COG2224#4 from 28211 to 1117',\n",
       "  'COG2241#2 from 28211 to 1117',\n",
       "  'COG2242#2 from 28211 to 1117',\n",
       "  'COG2243#3 from 28211 to 1117',\n",
       "  'COG2251#3 from 28211 to 1117',\n",
       "  'COG2263#3 from 28211 to 1117',\n",
       "  'COG2312#1 from 28211 to 1117',\n",
       "  'COG2321#2 from 28211 to 1117',\n",
       "  'COG2333#3 from 28211 to 1117',\n",
       "  'COG2336#4 from 28211 to 1117',\n",
       "  'COG2336#5 from 28211 to 1117',\n",
       "  'COG2345#4 from 28211 to 1117',\n",
       "  'COG2370#0 from 28211 to 1117',\n",
       "  'COG2371#1 from 28211 to 1117',\n",
       "  'COG2372#1 from 28211 to 1117',\n",
       "  'COG2372#1 from 28211 to 33071',\n",
       "  'COG2391#2 from 28211 to 1117',\n",
       "  'COG2461#0 from 28211 to 1117',\n",
       "  'COG2520#0 from 28211 to 1117',\n",
       "  'COG2608#3 from 28211 to 1117',\n",
       "  'COG2761#1 from 28211 to 1117',\n",
       "  'COG2823#0 from 28211 to 1117',\n",
       "  'COG2828#1 from 28211 to 1117',\n",
       "  'COG2855#7 from 28211 to 1117',\n",
       "  'COG2929#2 from 28211 to 1117',\n",
       "  'COG2929#2 from 356 to 1117',\n",
       "  'COG2932#0 from 28211 to 1117',\n",
       "  'COG2939#5 from 28211 to 1117',\n",
       "  'COG2944#8 from 28211 to 1117',\n",
       "  'COG2957#2 from 28211 to 1117',\n",
       "  'COG2994#0 from 28211 to 1117',\n",
       "  'COG3038#1 from 28211 to 1117',\n",
       "  'COG3128#2 from 28211 to 1117',\n",
       "  'COG3174#2 from 28211 to 1117',\n",
       "  'COG3188#1 from 28211 to 1117',\n",
       "  'COG3193#1 from 28211 to 1117',\n",
       "  'COG3221#3 from 28211 to 1117',\n",
       "  'COG3239#2 from 28211 to 1117',\n",
       "  'COG3251#0 from 356 to 1117',\n",
       "  'COG3271#7 from 28211 to 1117',\n",
       "  'COG3280#1 from 28211 to 1117',\n",
       "  'COG3306#0 from 28211 to 1117',\n",
       "  'COG3311#1 from 28211 to 1117',\n",
       "  'COG3349#2 from 28211 to 1117',\n",
       "  'COG3350#6 from 28211 to 1117',\n",
       "  'COG3379#0 from 28211 to 1117',\n",
       "  'COG3379#3 from 28211 to 1117',\n",
       "  'COG3439#1 from 28211 to 1117',\n",
       "  'COG3454#4 from 28211 to 1117',\n",
       "  'COG3510#0 from 28211 to 1117',\n",
       "  'COG3510#3 from 28211 to 1117',\n",
       "  'COG3514#1 from 28211 to 1117',\n",
       "  'COG3546#2 from 28211 to 1117',\n",
       "  'COG3551#2 from 28211 to 1117',\n",
       "  'COG3558#3 from 28211 to 1117',\n",
       "  'COG3571#3 from 28211 to 1117',\n",
       "  'COG3576#4 from 28211 to 1117',\n",
       "  'COG3576#5 from 28211 to 1117',\n",
       "  'COG3587#2 from 28211 to 1117',\n",
       "  'COG3608#0 from 28211 to 1117',\n",
       "  'COG3608#4 from 28211 to 1117',\n",
       "  'COG3609#5 from 28211 to 1117',\n",
       "  'COG3614#0 from 28211 to 1117',\n",
       "  'COG3619#1 from 28211 to 1117',\n",
       "  'COG3624#0 from 28211 to 1117',\n",
       "  'COG3625#1 from 28211 to 1117',\n",
       "  'COG3626#1 from 28211 to 1117',\n",
       "  'COG3636#7 from 28211 to 1117',\n",
       "  'COG3638#4 from 28211 to 1117',\n",
       "  'COG3657#0 from 28211 to 1117',\n",
       "  'COG3657#2 from 28211 to 1117',\n",
       "  'COG3665#1 from 28211 to 1117',\n",
       "  'COG3667#4 from 28211 to 1117',\n",
       "  'COG3673#3 from 28211 to 1117',\n",
       "  'COG3686#1 from 28211 to 1117',\n",
       "  'COG3703#1 from 28211 to 1117',\n",
       "  'COG3719#0 from 28211 to 1117',\n",
       "  'COG3743#0 from 28211 to 1117',\n",
       "  'COG3751#3 from 28211 to 1117',\n",
       "  'COG3754#1 from 28211 to 1117',\n",
       "  'COG3759#1 from 28211 to 1117',\n",
       "  'COG3788#0 from 28211 to 1117',\n",
       "  'COG3803#0 from 28211 to 1117',\n",
       "  'COG3805#3 from 204441 to 1117',\n",
       "  'COG3806#1 from 28211 to 1117',\n",
       "  'COG3807#0 from 28211 to 1117',\n",
       "  'COG3812#0 from 28211 to 1117',\n",
       "  'COG3837#2 from 28211 to 1117',\n",
       "  'COG3837#2 from 28211 to 33072',\n",
       "  'COG3839#0 from 28211 to 1117',\n",
       "  'COG3861#0 from 28211 to 1117',\n",
       "  'COG3861#0 from 28211 to 1416614',\n",
       "  'COG3865#1 from 28211 to 1117',\n",
       "  'COG3898#0 from 28211 to 1117',\n",
       "  'COG3932#1 from 28211 to 1117',\n",
       "  'COG3959#0 from 28211 to 1117',\n",
       "  'COG3971#0 from 28211 to 1117',\n",
       "  'COG4076#4 from 28211 to 1117',\n",
       "  'COG4113#2 from 28211 to 1117',\n",
       "  'COG4113#2 from 28211 to 1416614',\n",
       "  'COG4113#2 from 28211 to 33072',\n",
       "  'COG4126#3 from 28211 to 1117',\n",
       "  'COG4222#0 from 28211 to 1117',\n",
       "  'COG4222#0 from 28211 to 33072',\n",
       "  'COG4227#2 from 28211 to 1117',\n",
       "  'COG4232#1 from 28211 to 1117',\n",
       "  'COG4233#0 from 28211 to 1117',\n",
       "  'COG4244#4 from 28211 to 1117',\n",
       "  'COG4245#5 from 28211 to 1117',\n",
       "  'COG4249#5 from 28211 to 1117',\n",
       "  'COG4301#1 from 28211 to 1117',\n",
       "  'COG4310#1 from 356 to 1117',\n",
       "  'COG4319#9 from 28211 to 1117',\n",
       "  'COG4322#1 from 28211 to 1117',\n",
       "  'COG4325#1 from 28211 to 1117',\n",
       "  'COG4337#1 from 28211 to 1117',\n",
       "  'COG4338#0 from 28211 to 1117',\n",
       "  'COG4372#2 from 28211 to 1117',\n",
       "  'COG4373#1 from 28211 to 1117',\n",
       "  'COG4420#6 from 28211 to 1117',\n",
       "  'COG4453#0 from 28211 to 1117',\n",
       "  'COG4453#1 from 28211 to 1117',\n",
       "  'COG4454#2 from 28211 to 1117',\n",
       "  'COG4456#3 from 28211 to 1416614',\n",
       "  'COG4529#1 from 28211 to 1117',\n",
       "  'COG4576#4 from 41294 to 1117',\n",
       "  'COG4603#5 from 28211 to 1117',\n",
       "  'COG4603#6 from 28211 to 1117',\n",
       "  'COG4679#1 from 28211 to 1117',\n",
       "  'COG4683#0 from 28211 to 1117',\n",
       "  'COG4683#4 from 28211 to 1117',\n",
       "  'COG4683#4 from 28211 to 1416614',\n",
       "  'COG4716#1 from 28211 to 1117',\n",
       "  'COG4778#0 from 28211 to 1117',\n",
       "  'COG4803#1 from 28211 to 1117',\n",
       "  'COG4875#4 from 356 to 1117',\n",
       "  'COG4941#1 from 28211 to 1117',\n",
       "  'COG5266#4 from 28211 to 1117',\n",
       "  'COG5304#2 from 28211 to 1117',\n",
       "  'COG5309#5 from 28211 to 1117',\n",
       "  'COG5331#0 from 28211 to 1117',\n",
       "  'COG5350#0 from 28211 to 1117',\n",
       "  'COG5480#0 from 28211 to 1117',\n",
       "  'COG5499#3 from 28211 to 1117',\n",
       "  'COG5554#0 from 28211 to 1117',\n",
       "  'COG5554#1 from 41295 to 1117',\n",
       "  'COG5606#4 from 28211 to 1117',\n",
       "  'COG5606#4 from 356 to 1117',\n",
       "  'COG5622#0 from 28211 to 1117',\n",
       "  'COG5654#0 from 28211 to 33071',\n",
       "  'COG5662#0 from 28211 to 33072'},\n",
       " 'from 1117 to 1224': {'2Z7SC#2 from 1117 to 28211',\n",
       "  '2Z7SC#2 from 1117 to 356',\n",
       "  '2Z87Q#2 from 1117 to 356',\n",
       "  '2ZPQC#0 from 1117 to 356',\n",
       "  '31G5S#1 from 1117 to 517722',\n",
       "  '32SSN#0 from 1117 to 28211',\n",
       "  '32SSN#0 from 1117 to 425',\n",
       "  '32U2A#0 from 1117 to 1287116',\n",
       "  '32YJR#0 from 1117 to 28211',\n",
       "  '32YWC#0 from 1117 to 28211',\n",
       "  '3309S#1 from 1117 to 69657',\n",
       "  '330KH#0 from 1117 to 1217908',\n",
       "  '332XN#0 from 1117 to 28211',\n",
       "  '333NE#0 from 1117 to 31993',\n",
       "  '334XK#0 from 1117 to 28211',\n",
       "  'COG0001#0 from 1117 to 357',\n",
       "  'COG0119#3 from 1117 to 38323',\n",
       "  'COG0182#1 from 1117 to 1655514',\n",
       "  'COG0205#3 from 1117 to 28211',\n",
       "  'COG0344#1 from 1117 to 28211',\n",
       "  'COG0370#0 from 1117 to 28211',\n",
       "  'COG0378#3 from 1117 to 1046117',\n",
       "  'COG0378#3 from 1117 to 28211',\n",
       "  'COG0398#1 from 1117 to 1124597',\n",
       "  'COG0398#1 from 1117 to 204799',\n",
       "  'COG0398#1 from 1117 to 28211',\n",
       "  'COG0398#1 from 1117 to 288022',\n",
       "  'COG0398#1 from 1117 to 523',\n",
       "  'COG0401#0 from 1117 to 28211',\n",
       "  'COG0401#0 from 1117 to 397260',\n",
       "  'COG0529#1 from 1117 to 28211',\n",
       "  'COG0545#3 from 1117 to 1777752',\n",
       "  'COG0545#3 from 1117 to 28211',\n",
       "  'COG0563#5 from 1117 to 64973',\n",
       "  'COG0568#1 from 1117 to 28211',\n",
       "  'COG0581#5 from 1117 to 356',\n",
       "  'COG0608#1 from 1117 to 391593',\n",
       "  'COG0618#1 from 1117 to 28211',\n",
       "  'COG0636#7 from 1117 to 187304',\n",
       "  'COG0636#7 from 1117 to 985054',\n",
       "  'COG0639#4 from 1117 to 28211',\n",
       "  'COG0639#4 from 1117 to 356',\n",
       "  'COG0668#5 from 1117 to 6',\n",
       "  'COG0678#2 from 1117 to 28211',\n",
       "  'COG0678#2 from 1117 to 69657',\n",
       "  'COG0689#1 from 1117 to 204455',\n",
       "  'COG0689#1 from 1117 to 28211',\n",
       "  'COG0695#3 from 1117 to 28211',\n",
       "  'COG0728#1 from 1117 to 334545',\n",
       "  'COG0757#3 from 1117 to 28211',\n",
       "  'COG0783#0 from 1117 to 195105',\n",
       "  'COG0831#2 from 1117 to 204441',\n",
       "  'COG1262#7 from 1117 to 1069',\n",
       "  'COG1262#7 from 1117 to 1124597',\n",
       "  'COG1262#7 from 1117 to 1380355',\n",
       "  'COG1262#7 from 1117 to 379',\n",
       "  'COG1285#2 from 1117 to 28211',\n",
       "  'COG1301#14 from 1117 to 28211',\n",
       "  'COG1393#0 from 1117 to 542',\n",
       "  'COG1524#2 from 1117 to 374',\n",
       "  'COG1598#2 from 1117 to 1502724',\n",
       "  'COG1633#10 from 1117 to 433',\n",
       "  'COG1661#2 from 1117 to 28211',\n",
       "  'COG1830#4 from 1117 to 663610',\n",
       "  'COG1898#2 from 1117 to 1124597',\n",
       "  'COG1898#2 from 1117 to 1655514',\n",
       "  'COG1898#2 from 1117 to 204457',\n",
       "  'COG1918#0 from 1117 to 28211',\n",
       "  'COG2109#2 from 1117 to 28211',\n",
       "  'COG2109#2 from 1117 to 41297',\n",
       "  'COG2223#4 from 1117 to 28211',\n",
       "  'COG2236#4 from 1117 to 1655514',\n",
       "  'COG2236#4 from 1117 to 198252',\n",
       "  'COG2327#5 from 1117 to 1320556',\n",
       "  'COG2336#0 from 1117 to 204441',\n",
       "  'COG2336#0 from 1117 to 204455',\n",
       "  'COG2336#0 from 1117 to 28211',\n",
       "  'COG2336#0 from 1117 to 41297',\n",
       "  'COG2336#0 from 1117 to 76890',\n",
       "  'COG2336#0 from 1117 to 76892',\n",
       "  'COG2336#0 from 1117 to 85',\n",
       "  'COG2340#1 from 1117 to 28211',\n",
       "  'COG2370#0 from 1117 to 28211',\n",
       "  'COG2370#0 from 1117 to 365533',\n",
       "  'COG2370#0 from 1117 to 531813',\n",
       "  'COG2403#1 from 1117 to 457934',\n",
       "  'COG2404#4 from 1117 to 28211',\n",
       "  'COG2461#0 from 1117 to 1379701',\n",
       "  'COG2764#1 from 1117 to 28211',\n",
       "  'COG2764#1 from 1117 to 41294',\n",
       "  'COG2823#0 from 1117 to 172088',\n",
       "  'COG2831#3 from 1117 to 42443',\n",
       "  'COG3038#2 from 1117 to 743836',\n",
       "  'COG3126#0 from 1117 to 1333998',\n",
       "  'COG3206#7 from 1117 to 1040987',\n",
       "  'COG3206#7 from 1117 to 28211',\n",
       "  'COG3221#5 from 1117 to 204441',\n",
       "  'COG3221#5 from 1117 to 28211',\n",
       "  'COG3267#3 from 1117 to 1035',\n",
       "  'COG3267#3 from 1117 to 28211',\n",
       "  'COG3411#8 from 1117 to 204457',\n",
       "  'COG3459#6 from 1117 to 28211',\n",
       "  'COG3514#1 from 1117 to 356',\n",
       "  'COG3521#0 from 1117 to 981384',\n",
       "  'COG3546#2 from 1117 to 716925',\n",
       "  'COG3550#1 from 1117 to 334852',\n",
       "  'COG3657#2 from 1117 to 107636',\n",
       "  'COG3657#2 from 1117 to 356',\n",
       "  'COG3657#2 from 1117 to 44746',\n",
       "  'COG3657#2 from 1117 to 774',\n",
       "  'COG3695#1 from 1117 to 204441',\n",
       "  'COG3719#0 from 1117 to 77038',\n",
       "  'COG3754#1 from 1117 to 263378',\n",
       "  'COG3754#1 from 1117 to 28211',\n",
       "  'COG3754#1 from 1117 to 373',\n",
       "  'COG3788#1 from 1117 to 204441',\n",
       "  'COG3793#0 from 1117 to 28211',\n",
       "  'COG3793#0 from 1117 to 85',\n",
       "  'COG3803#2 from 1117 to 28211',\n",
       "  'COG3805#0 from 1117 to 204457',\n",
       "  'COG3895#0 from 1117 to 204441',\n",
       "  'COG3898#0 from 1117 to 28211',\n",
       "  'COG3965#2 from 1117 to 28211',\n",
       "  'COG4076#0 from 1117 to 1248917',\n",
       "  'COG4076#0 from 1117 to 81',\n",
       "  'COG4101#0 from 1117 to 133',\n",
       "  'COG4102#0 from 1117 to 28211',\n",
       "  'COG4102#0 from 1117 to 357',\n",
       "  'COG4113#0 from 1117 to 208216',\n",
       "  'COG4113#0 from 1117 to 28211',\n",
       "  'COG4176#2 from 1117 to 28211',\n",
       "  'COG4178#2 from 1117 to 82115',\n",
       "  'COG4320#2 from 1117 to 204441',\n",
       "  'COG4374#1 from 1117 to 28211',\n",
       "  'COG4402#2 from 1117 to 28211',\n",
       "  'COG4421#3 from 1117 to 198252',\n",
       "  'COG4421#3 from 1117 to 28211',\n",
       "  'COG4453#0 from 1117 to 28211',\n",
       "  'COG4453#0 from 1117 to 426117',\n",
       "  'COG4453#1 from 1117 to 284016',\n",
       "  'COG4453#1 from 1117 to 82115',\n",
       "  'COG4453#3 from 1117 to 1076',\n",
       "  'COG4453#3 from 1117 to 1187851',\n",
       "  'COG4453#3 from 1117 to 28211',\n",
       "  'COG4453#3 from 1117 to 356',\n",
       "  'COG4454#2 from 1117 to 41294',\n",
       "  'COG4467#0 from 1117 to 1087',\n",
       "  'COG4678#1 from 1117 to 28211',\n",
       "  'COG4714#2 from 1117 to 28211',\n",
       "  'COG5018#1 from 1117 to 1282361',\n",
       "  'COG5031#4 from 1117 to 28211',\n",
       "  'COG5266#2 from 1117 to 245188',\n",
       "  'COG5293#1 from 1117 to 76745',\n",
       "  'COG5361#5 from 1117 to 31989',\n",
       "  'COG5381#2 from 1117 to 41295',\n",
       "  'COG5428#0 from 1117 to 28211',\n",
       "  'COG5439#1 from 1117 to 41295',\n",
       "  'COG5606#4 from 1117 to 28211',\n",
       "  'COG5606#4 from 1117 to 356',\n",
       "  'COG5606#4 from 1117 to 56946',\n",
       "  'COG5625#0 from 1117 to 1084',\n",
       "  'COG5625#0 from 1117 to 28211',\n",
       "  'COG5625#0 from 1117 to 356',\n",
       "  'COG5662#0 from 1117 to 1121829'},\n",
       " 'sisters': {'2Z7NV#1',\n",
       "  '2Z9EI#0',\n",
       "  '2Z9EI#1',\n",
       "  '2ZA6T#2',\n",
       "  '30VW1#2',\n",
       "  '32WB4#0',\n",
       "  '32YQJ#0',\n",
       "  '32Z2Q#2',\n",
       "  '32ZHQ#0',\n",
       "  '32ZWX#0',\n",
       "  '330KH#0',\n",
       "  '331H5#1',\n",
       "  '334XK#0',\n",
       "  '33B8C#0',\n",
       "  '33BC0#0',\n",
       "  'COG0001#0',\n",
       "  'COG0057#4',\n",
       "  'COG0308#5',\n",
       "  'COG0378#3',\n",
       "  'COG0398#1',\n",
       "  'COG0401#0',\n",
       "  'COG0489#5',\n",
       "  'COG0501#1',\n",
       "  'COG0568#1',\n",
       "  'COG0591#6',\n",
       "  'COG0618#1',\n",
       "  'COG0645#5',\n",
       "  'COG0678#2',\n",
       "  'COG0689#1',\n",
       "  'COG0694#2',\n",
       "  'COG0695#4',\n",
       "  'COG0709#3',\n",
       "  'COG0728#1',\n",
       "  'COG0786#4',\n",
       "  'COG0819#1',\n",
       "  'COG0831#2',\n",
       "  'COG1045#5',\n",
       "  'COG1204#5',\n",
       "  'COG1230#0',\n",
       "  'COG1301#14',\n",
       "  'COG1317#2',\n",
       "  'COG1322#6',\n",
       "  'COG1363#0',\n",
       "  'COG1390#4',\n",
       "  'COG1598#1',\n",
       "  'COG1682#8',\n",
       "  'COG1740#3',\n",
       "  'COG1830#4',\n",
       "  'COG1876#4',\n",
       "  'COG1994#5',\n",
       "  'COG2002#8',\n",
       "  'COG2073#2',\n",
       "  'COG2126#0',\n",
       "  'COG2246#1',\n",
       "  'COG2370#0',\n",
       "  'COG2391#5',\n",
       "  'COG2445#6',\n",
       "  'COG2520#0',\n",
       "  'COG2770#1',\n",
       "  'COG2906#3',\n",
       "  'COG2932#0',\n",
       "  'COG3019#0',\n",
       "  'COG3111#5',\n",
       "  'COG3128#2',\n",
       "  'COG3287#0',\n",
       "  'COG3306#7',\n",
       "  'COG3326#0',\n",
       "  'COG3379#0',\n",
       "  'COG3379#3',\n",
       "  'COG3435#0',\n",
       "  'COG3447#0',\n",
       "  'COG3491#3',\n",
       "  'COG3510#3',\n",
       "  'COG3609#6',\n",
       "  'COG3615#0',\n",
       "  'COG3657#0',\n",
       "  'COG3686#1',\n",
       "  'COG3807#0',\n",
       "  'COG3898#0',\n",
       "  'COG3950#5',\n",
       "  'COG4067#0',\n",
       "  'COG4076#4',\n",
       "  'COG4101#0',\n",
       "  'COG4107#1',\n",
       "  'COG4113#2',\n",
       "  'COG4176#1',\n",
       "  'COG4297#1',\n",
       "  'COG4301#1',\n",
       "  'COG4320#2',\n",
       "  'COG4322#1',\n",
       "  'COG4374#1',\n",
       "  'COG4453#0',\n",
       "  'COG4454#2',\n",
       "  'COG4634#4',\n",
       "  'COG4691#3',\n",
       "  'COG4803#1',\n",
       "  'COG4935#5',\n",
       "  'COG5007#0',\n",
       "  'COG5450#2',\n",
       "  'COG5463#0',\n",
       "  'COG5526#1',\n",
       "  'COG5554#0',\n",
       "  'COG5654#0'},\n",
       " 'no clear relationship between phyla': {'2Z7KB#0',\n",
       "  '2Z7N0#0',\n",
       "  '2Z7N9#1',\n",
       "  '2Z7NV#3',\n",
       "  '2Z7NZ#1',\n",
       "  '2Z7Q6#2',\n",
       "  '2Z7WD#3',\n",
       "  '2Z80X#3',\n",
       "  '2Z854#0',\n",
       "  '2Z87F#2',\n",
       "  '2Z8IF#0',\n",
       "  '2Z8T4#1',\n",
       "  '2Z8T5#0',\n",
       "  '2Z8UG#1',\n",
       "  '2Z8UT#0',\n",
       "  '2Z8XI#0',\n",
       "  '2Z93N#0',\n",
       "  '2Z976#0',\n",
       "  '2Z98N#4',\n",
       "  '2Z9V7#1',\n",
       "  '2Z9VT#0',\n",
       "  '2Z9YC#0',\n",
       "  '2ZA6T#1',\n",
       "  '2ZAAA#2',\n",
       "  '2ZB6Z#2',\n",
       "  '2ZC7Q#3',\n",
       "  '2ZCFZ#0',\n",
       "  '2ZCI8#0',\n",
       "  '2ZR3M#0',\n",
       "  '30BIC#1',\n",
       "  '30DUI#0',\n",
       "  '30H5P#0',\n",
       "  '31MME#0',\n",
       "  '31NK6#0',\n",
       "  '31YEN#1',\n",
       "  '326SN#0',\n",
       "  '32EPT#0',\n",
       "  '32NM5#2',\n",
       "  '32RWS#1',\n",
       "  '32RYR#0',\n",
       "  '32S87#1',\n",
       "  '32SI8#0',\n",
       "  '32VVI#0',\n",
       "  '32VXZ#3',\n",
       "  '32XNF#0',\n",
       "  '32YAP#4',\n",
       "  '32YBX#0',\n",
       "  '32YHH#1',\n",
       "  '32YN6#0',\n",
       "  '32YT2#0',\n",
       "  '32YVY#2',\n",
       "  '32ZB4#0',\n",
       "  '32ZUE#0',\n",
       "  '3302Q#0',\n",
       "  '330AH#2',\n",
       "  '330DP#2',\n",
       "  '330SE#0',\n",
       "  '333DW#2',\n",
       "  '336QS#0',\n",
       "  '33AHD#0',\n",
       "  '33AKF#0',\n",
       "  '33E3H#0',\n",
       "  'COG0003#1',\n",
       "  'COG0006#3',\n",
       "  'COG0006#4',\n",
       "  'COG0008#0',\n",
       "  'COG0009#1',\n",
       "  'COG0011#2',\n",
       "  'COG0023#0',\n",
       "  'COG0024#2',\n",
       "  'COG0027#4',\n",
       "  'COG0033#2',\n",
       "  'COG0038#8',\n",
       "  'COG0048#0',\n",
       "  'COG0053#2',\n",
       "  'COG0069#0',\n",
       "  'COG0094#2',\n",
       "  'COG0121#2',\n",
       "  'COG0121#4',\n",
       "  'COG0144#0',\n",
       "  'COG0165#0',\n",
       "  'COG0220#0',\n",
       "  'COG0227#2',\n",
       "  'COG0229#2',\n",
       "  'COG0242#0',\n",
       "  'COG0246#1',\n",
       "  'COG0247#6',\n",
       "  'COG0252#3',\n",
       "  'COG0272#7',\n",
       "  'COG0278#0',\n",
       "  'COG0285#5',\n",
       "  'COG0287#7',\n",
       "  'COG0298#2',\n",
       "  'COG0303#3',\n",
       "  'COG0306#2',\n",
       "  'COG0310#0',\n",
       "  'COG0314#2',\n",
       "  'COG0314#3',\n",
       "  'COG0317#9',\n",
       "  'COG0322#6',\n",
       "  'COG0339#0',\n",
       "  'COG0340#0',\n",
       "  'COG0349#5',\n",
       "  'COG0351#1',\n",
       "  'COG0369#4',\n",
       "  'COG0370#2',\n",
       "  'COG0372#0',\n",
       "  'COG0373#0',\n",
       "  'COG0374#1',\n",
       "  'COG0382#3',\n",
       "  'COG0387#4',\n",
       "  'COG0390#2',\n",
       "  'COG0392#2',\n",
       "  'COG0393#2',\n",
       "  'COG0396#3',\n",
       "  'COG0397#3',\n",
       "  'COG0397#5',\n",
       "  'COG0398#3',\n",
       "  'COG0407#3',\n",
       "  'COG0418#1',\n",
       "  'COG0430#1',\n",
       "  'COG0439#3',\n",
       "  'COG0441#6',\n",
       "  'COG0467#5',\n",
       "  'COG0474#3',\n",
       "  'COG0480#6',\n",
       "  'COG0493#2',\n",
       "  'COG0499#5',\n",
       "  'COG0507#9',\n",
       "  'COG0510#0',\n",
       "  'COG0512#1',\n",
       "  'COG0514#3',\n",
       "  'COG0522#0',\n",
       "  'COG0533#3',\n",
       "  'COG0535#1',\n",
       "  'COG0540#1',\n",
       "  'COG0552#1',\n",
       "  'COG0557#1',\n",
       "  'COG0560#2',\n",
       "  'COG0569#1',\n",
       "  'COG0573#0',\n",
       "  'COG0579#2',\n",
       "  'COG0591#8',\n",
       "  'COG0594#0',\n",
       "  'COG0627#2',\n",
       "  'COG0628#12',\n",
       "  'COG0630#3',\n",
       "  'COG0655#3',\n",
       "  'COG0663#1',\n",
       "  'COG0668#11',\n",
       "  'COG0668#2',\n",
       "  'COG0670#6',\n",
       "  'COG0676#0',\n",
       "  'COG0678#0',\n",
       "  'COG0679#4',\n",
       "  'COG0701#5',\n",
       "  'COG0716#4',\n",
       "  'COG0719#4',\n",
       "  'COG0720#0',\n",
       "  'COG0722#1',\n",
       "  'COG0723#1',\n",
       "  'COG0755#0',\n",
       "  'COG0757#0',\n",
       "  'COG0758#3',\n",
       "  'COG0759#5',\n",
       "  'COG0769#1',\n",
       "  'COG0785#4',\n",
       "  'COG0792#0',\n",
       "  'COG0798#0',\n",
       "  'COG0800#1',\n",
       "  'COG0804#2',\n",
       "  'COG0824#7',\n",
       "  'COG0826#8',\n",
       "  'COG0828#0',\n",
       "  'COG0829#3',\n",
       "  'COG0830#3',\n",
       "  'COG0847#2',\n",
       "  'COG0854#2',\n",
       "  'COG0857#1',\n",
       "  'COG0863#3',\n",
       "  'COG0864#1',\n",
       "  'COG1010#1',\n",
       "  'COG1067#1',\n",
       "  'COG1076#0',\n",
       "  'COG1079#6',\n",
       "  'COG1100#5',\n",
       "  'COG1117#4',\n",
       "  'COG1124#0',\n",
       "  'COG1134#2',\n",
       "  'COG1141#0',\n",
       "  'COG1145#2',\n",
       "  'COG1149#3',\n",
       "  'COG1155#1',\n",
       "  'COG1171#1',\n",
       "  'COG1201#2',\n",
       "  'COG1218#4',\n",
       "  'COG1222#1',\n",
       "  'COG1226#3',\n",
       "  'COG1230#2',\n",
       "  'COG1232#2',\n",
       "  'COG1234#5',\n",
       "  'COG1237#0',\n",
       "  'COG1251#4',\n",
       "  'COG1262#8',\n",
       "  'COG1269#5',\n",
       "  'COG1282#0',\n",
       "  'COG1283#1',\n",
       "  'COG1293#1',\n",
       "  'COG1294#0',\n",
       "  'COG1301#12',\n",
       "  'COG1305#4',\n",
       "  'COG1330#1',\n",
       "  'COG1336#3',\n",
       "  'COG1340#3',\n",
       "  'COG1359#2',\n",
       "  'COG1363#2',\n",
       "  'COG1384#3',\n",
       "  'COG1385#6',\n",
       "  'COG1392#0',\n",
       "  'COG1394#1',\n",
       "  'COG1402#2',\n",
       "  'COG1446#2',\n",
       "  'COG1449#2',\n",
       "  'COG1449#6',\n",
       "  'COG1457#17',\n",
       "  'COG1474#5',\n",
       "  'COG1487#3',\n",
       "  'COG1496#1',\n",
       "  'COG1515#3',\n",
       "  'COG1528#7',\n",
       "  'COG1573#2',\n",
       "  'COG1626#1',\n",
       "  'COG1628#1',\n",
       "  'COG1640#0',\n",
       "  'COG1643#2',\n",
       "  'COG1657#2',\n",
       "  'COG1671#2',\n",
       "  'COG1682#4',\n",
       "  'COG1682#5',\n",
       "  'COG1690#1',\n",
       "  'COG1700#2',\n",
       "  'COG1704#1',\n",
       "  'COG1720#2',\n",
       "  'COG1721#1',\n",
       "  'COG1724#4',\n",
       "  'COG1741#1',\n",
       "  'COG1748#0',\n",
       "  'COG1752#3',\n",
       "  'COG1776#5',\n",
       "  'COG1783#3',\n",
       "  'COG1791#1',\n",
       "  'COG1820#2',\n",
       "  'COG1839#0',\n",
       "  'COG1842#8',\n",
       "  'COG1850#0',\n",
       "  'COG1850#3',\n",
       "  'COG1857#1',\n",
       "  'COG1858#3',\n",
       "  'COG1875#0',\n",
       "  'COG1895#2',\n",
       "  'COG1899#3',\n",
       "  'COG1914#3',\n",
       "  'COG1922#3',\n",
       "  'COG1926#0',\n",
       "  'COG1928#1',\n",
       "  'COG1948#0',\n",
       "  'COG1959#1',\n",
       "  'COG1971#5',\n",
       "  'COG1984#2',\n",
       "  'COG1988#0',\n",
       "  'COG1994#6',\n",
       "  'COG2002#3',\n",
       "  'COG2014#3',\n",
       "  'COG2041#4',\n",
       "  'COG2046#3',\n",
       "  'COG2049#1',\n",
       "  'COG2059#11',\n",
       "  'COG2070#1',\n",
       "  'COG2087#2',\n",
       "  'COG2100#1',\n",
       "  'COG2140#1',\n",
       "  'COG2144#1',\n",
       "  'COG2146#3',\n",
       "  'COG2162#3',\n",
       "  'COG2170#5',\n",
       "  'COG2193#2',\n",
       "  'COG2193#3',\n",
       "  'COG2193#5',\n",
       "  'COG2193#6',\n",
       "  'COG2194#2',\n",
       "  'COG2220#4',\n",
       "  'COG2233#4',\n",
       "  'COG2235#3',\n",
       "  'COG2251#2',\n",
       "  'COG2268#4',\n",
       "  'COG2314#1',\n",
       "  'COG2329#9',\n",
       "  'COG2334#4',\n",
       "  'COG2336#2',\n",
       "  'COG2371#2',\n",
       "  'COG2410#1',\n",
       "  'COG2421#3',\n",
       "  'COG2433#1',\n",
       "  'COG2509#3',\n",
       "  'COG2516#2',\n",
       "  'COG2603#1',\n",
       "  'COG2718#0',\n",
       "  'COG2740#3',\n",
       "  'COG2821#1',\n",
       "  'COG2848#0',\n",
       "  'COG2876#1',\n",
       "  'COG2877#1',\n",
       "  'COG2886#1',\n",
       "  'COG2886#3',\n",
       "  'COG2886#4',\n",
       "  'COG2886#6',\n",
       "  'COG2888#0',\n",
       "  'COG2905#2',\n",
       "  'COG2906#0',\n",
       "  'COG2930#0',\n",
       "  'COG2941#2',\n",
       "  'COG2954#1',\n",
       "  'COG2979#0',\n",
       "  'COG2988#2',\n",
       "  'COG3019#3',\n",
       "  'COG3021#2',\n",
       "  'COG3040#1',\n",
       "  'COG3041#0',\n",
       "  'COG3105#1',\n",
       "  'COG3135#5',\n",
       "  'COG3146#2',\n",
       "  'COG3161#1',\n",
       "  'COG3168#5',\n",
       "  'COG3176#4',\n",
       "  'COG3182#6',\n",
       "  'COG3214#1',\n",
       "  'COG3219#0',\n",
       "  'COG3220#4',\n",
       "  'COG3224#0',\n",
       "  'COG3225#3',\n",
       "  'COG3227#3',\n",
       "  'COG3239#3',\n",
       "  'COG3247#2',\n",
       "  'COG3254#5',\n",
       "  'COG3256#4',\n",
       "  'COG3268#3',\n",
       "  'COG3288#2',\n",
       "  'COG3304#2',\n",
       "  'COG3329#3',\n",
       "  'COG3395#0',\n",
       "  'COG3396#8',\n",
       "  'COG3408#9',\n",
       "  'COG3424#4',\n",
       "  'COG3429#0',\n",
       "  'COG3431#1',\n",
       "  'COG3431#2',\n",
       "  'COG3448#6',\n",
       "  'COG3449#4',\n",
       "  'COG3450#0',\n",
       "  'COG3454#0',\n",
       "  'COG3471#1',\n",
       "  'COG3484#0',\n",
       "  'COG3515#0',\n",
       "  'COG3524#2',\n",
       "  'COG3542#1',\n",
       "  'COG3546#4',\n",
       "  'COG3562#1',\n",
       "  'COG3563#1',\n",
       "  'COG3565#1',\n",
       "  'COG3585#0',\n",
       "  'COG3585#2',\n",
       "  'COG3591#1',\n",
       "  'COG3607#3',\n",
       "  'COG3608#5',\n",
       "  'COG3614#1',\n",
       "  'COG3624#1',\n",
       "  'COG3626#2',\n",
       "  'COG3627#1',\n",
       "  'COG3636#3',\n",
       "  'COG3639#4',\n",
       "  'COG3639#7',\n",
       "  'COG3640#2',\n",
       "  'COG3653#4',\n",
       "  'COG3663#1',\n",
       "  'COG3665#0',\n",
       "  'COG3670#4',\n",
       "  'COG3675#4',\n",
       "  'COG3686#3',\n",
       "  'COG3719#2',\n",
       "  'COG3729#1',\n",
       "  'COG3743#6',\n",
       "  'COG3744#4',\n",
       "  'COG3751#2',\n",
       "  'COG3751#5',\n",
       "  'COG3752#4',\n",
       "  'COG3755#4',\n",
       "  'COG3769#0',\n",
       "  'COG3772#3',\n",
       "  'COG3794#4',\n",
       "  'COG3812#1',\n",
       "  'COG3854#1',\n",
       "  'COG3900#0',\n",
       "  'COG3950#2',\n",
       "  'COG3954#0',\n",
       "  'COG3957#2',\n",
       "  'COG3961#3',\n",
       "  'COG3965#0',\n",
       "  'COG3968#0',\n",
       "  'COG3975#5',\n",
       "  'COG3980#3',\n",
       "  'COG4097#2',\n",
       "  'COG4103#0',\n",
       "  'COG4113#6',\n",
       "  'COG4114#2',\n",
       "  'COG4115#3',\n",
       "  'COG4158#6',\n",
       "  'COG4175#3',\n",
       "  'COG4222#4',\n",
       "  'COG4226#2',\n",
       "  'COG4243#3',\n",
       "  'COG4245#3',\n",
       "  'COG4248#2',\n",
       "  'COG4276#2',\n",
       "  'COG4278#4',\n",
       "  'COG4279#4',\n",
       "  'COG4291#3',\n",
       "  'COG4298#0',\n",
       "  'COG4319#13',\n",
       "  'COG4325#2',\n",
       "  'COG4328#3',\n",
       "  'COG4329#1',\n",
       "  'COG4337#3',\n",
       "  'COG4340#2',\n",
       "  'COG4359#4',\n",
       "  'COG4360#3',\n",
       "  'COG4392#0',\n",
       "  'COG4445#0',\n",
       "  'COG4451#2',\n",
       "  'COG4456#5',\n",
       "  'COG4531#2',\n",
       "  'COG4551#0',\n",
       "  'COG4552#0',\n",
       "  'COG4570#0',\n",
       "  'COG4574#0',\n",
       "  'COG4632#5',\n",
       "  'COG4633#3',\n",
       "  'COG4634#1',\n",
       "  'COG4665#1',\n",
       "  'COG4671#0',\n",
       "  'COG4671#1',\n",
       "  'COG4680#0',\n",
       "  'COG4710#0',\n",
       "  'COG4735#0',\n",
       "  'COG4798#3',\n",
       "  'COG4823#3',\n",
       "  'COG4861#5',\n",
       "  'COG4902#2',\n",
       "  'COG4930#0',\n",
       "  'COG4938#2',\n",
       "  'COG4968#1',\n",
       "  'COG5135#2',\n",
       "  'COG5184#1',\n",
       "  'COG5267#2',\n",
       "  'COG5278#1',\n",
       "  'COG5278#2',\n",
       "  'COG5304#0',\n",
       "  'COG5305#3',\n",
       "  'COG5316#3',\n",
       "  'COG5340#4',\n",
       "  'COG5378#1',\n",
       "  'COG5416#6',\n",
       "  'COG5428#5',\n",
       "  'COG5435#0',\n",
       "  'COG5450#3',\n",
       "  'COG5483#2',\n",
       "  'COG5493#1',\n",
       "  'COG5493#2',\n",
       "  'COG5553#1',\n",
       "  'COG5563#2',\n",
       "  'COG5606#2',\n",
       "  'COG5610#0',\n",
       "  'COG5611#2',\n",
       "  'COG5626#1',\n",
       "  'COG5634#1',\n",
       "  'COG5634#4',\n",
       "  'COG5635#2'}}"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assess_transfers_between_phyla((1224, 1117))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1090, 1224)\n",
      "(1117, 1224)\n",
      "(1224, 200795)\n",
      "(1224, 976)\n",
      "(1224, 1134404)\n",
      "(1224, 1798710)\n"
     ]
    }
   ],
   "source": [
    "descriptions = {}\n",
    "for phylum_pair in itertools.combinations(sampled_phyla, 2):\n",
    "# for phylum_pair in [(200795, 1117)]:\n",
    "    if 1224 not in phylum_pair:\n",
    "        continue\n",
    "    \n",
    "    print(phylum_pair)\n",
    "    descriptions[phylum_pair] = assess_transfers_between_phyla(phylum_pair)\n",
    "    \n",
    "# descriptions = assess_transfers_between_phyla((200795, 1117))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1090-1224\n",
      "   from 1090 to 1224: 0\n",
      "   from 1224 to 1090: 39\n",
      "   sisters: 6\n",
      "   no clear relationship between phyla: 180\n",
      "\n",
      "1117-1224\n",
      "   from 1117 to 1224: 164\n",
      "   from 1224 to 1117: 279\n",
      "   sisters: 103\n",
      "   no clear relationship between phyla: 486\n",
      "\n",
      "1224-200795\n",
      "   from 1224 to 200795: 23\n",
      "   from 200795 to 1224: 10\n",
      "   sisters: 8\n",
      "   no clear relationship between phyla: 87\n",
      "\n",
      "1224-976\n",
      "   from 1224 to 976: 307\n",
      "   from 976 to 1224: 338\n",
      "   sisters: 106\n",
      "   no clear relationship between phyla: 777\n",
      "\n",
      "1224-1134404\n",
      "   from 1224 to 1134404: 0\n",
      "   from 1134404 to 1224: 0\n",
      "   sisters: 0\n",
      "   no clear relationship between phyla: 0\n",
      "\n",
      "1224-1798710\n",
      "   from 1224 to 1798710: 0\n",
      "   from 1798710 to 1224: 0\n",
      "   sisters: 0\n",
      "   no clear relationship between phyla: 0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "candidates_to_improve = set()\n",
    "for pair, description in descriptions.items():\n",
    "#     if 1224 in pair:\n",
    "#         continue\n",
    "        \n",
    "    print('%i-%i' % pair)\n",
    "    for key, value in description.items():\n",
    "        if re.match('from \\d+ to \\d+$', key):\n",
    "            for candidate in value:\n",
    "                candidates_to_improve.add(candidate.split()[0])\n",
    "            \n",
    "        print('  ', key+':', len(value))\n",
    "    \n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for candidate in candidates_to_improve:\n",
    "    group_id, cluster_num = candidate.split('#')\n",
    "    \n",
    "    if not os.path.isfile(f'candidates/alignments/{group_id}-cluster{cluster_num}.aln'):\n",
    "        print(candidate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('candidates/candidates_to_run_iqtree', 'w') as out:\n",
    "    \n",
    "    for candidate in candidates_to_improve:\n",
    "        group_id, cluster_num = candidate.split('#')\n",
    "        out.write(f'{group_id}-cluster{cluster_num}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('\\n'.join(descriptions[(1117,1224)]['from 1224 to 1117']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "related_phyla = (200795, 1117)\n",
    "descriptions  = {f'from {related_phyla[0]} to {related_phyla[1]}': set(),\n",
    "                 f'from {related_phyla[1]} to {related_phyla[0]}': set(),\n",
    "                  'sisters':                                       set(),\n",
    "                  'no clear relationship between phyla':           set()}\n",
    "candidates    = set(significant_df.query('source==\"%i\" and target==\"%i\"' % related_phyla).group.values).union(\n",
    "                    significant_df.query('target==\"%i\" and source==\"%i\"' % related_phyla).group.values\n",
    "                )\n",
    "\n",
    "for candidate in candidates:\n",
    "\n",
    "    group_id, cluster_num = candidate.split('#')\n",
    "\n",
    "    with cd('candidates/trees/'):\n",
    "        #\n",
    "        # test if there is rooted version of the tree (mad output) \n",
    "        if not os.path.isfile('%s-cluster%s.fastTree.rooted' % (group_id, cluster_num)):\n",
    "            continue\n",
    "\n",
    "        #\n",
    "        # we need the original tree cause mad removes support values, so we just transfer root positions\n",
    "        try:\n",
    "            tmp_tree   = match_rooting(ete3.Tree('%s-cluster%s.fastTree.rooted' % (group_id, cluster_num)),\n",
    "                                       ete3.Tree('%s-cluster%s.fastTree'        % (group_id, cluster_num)))\n",
    "        except ete3.parser.newick.NewickError:\n",
    "            continue\n",
    "\n",
    "    #\n",
    "    # extract taxIDs to subsample taxonomy tables\n",
    "    taxids = {}\n",
    "    for leaf in tmp_tree.get_leaf_names():\n",
    "        #\n",
    "        # leaf names are composed by <taxid>.<locus_tag>\n",
    "        #   ps: locus_tag may also have \".\" within it\n",
    "        tmp_taxid = int(leaf.split('.')[0])\n",
    "\n",
    "        if tmp_taxid not in taxids:\n",
    "            taxids[tmp_taxid] = []\n",
    "        taxids[tmp_taxid].append(leaf)\n",
    "\n",
    "    #\n",
    "    # select taxIDs from each assessed phylum...\n",
    "    phylum1_taxonomy = eggNOG_taxonomy.loc[taxids].query('phylum==@related_phyla[0]')\n",
    "    phylum2_taxonomy = eggNOG_taxonomy.loc[taxids].query('phylum==@related_phyla[1]')\n",
    "\n",
    "    #\n",
    "    # ... and their respect leaves\n",
    "    phylum1_leaves = set()\n",
    "    for taxid in phylum1_taxonomy.index:\n",
    "        phylum1_leaves.update(taxids[taxid])\n",
    "\n",
    "    phylum2_leaves = set()\n",
    "    for taxid in phylum2_taxonomy.index:\n",
    "        phylum2_leaves.update(taxids[taxid])\n",
    "\n",
    "    all_leaves = phylum1_leaves.union(phylum2_leaves)\n",
    "\n",
    "    #\n",
    "    # as tree traversing through ete3 isn't very efficient, and doesn't scalate very well,\n",
    "    #   we create an iGraph dag for more effient traversing\n",
    "    dag  = tree_to_dag(tmp_tree)          # we need a directed version...\n",
    "    udag = dag.as_undirected(mode='each') # ... and an undirected one for different processes\n",
    "\n",
    "    #\n",
    "    # placeholders where we will add monophyletic clades for each phylum\n",
    "    phylum1_clades = set() \n",
    "    phylum2_clades = set()\n",
    "\n",
    "    ignored_nodes  = [] # descendants of monophyletic nodes should be ignored once their\n",
    "                        #   ancestors have been stored.\n",
    "\n",
    "    #\n",
    "    # traverse through internal nodes\n",
    "    for node in dag.vs:\n",
    "        if node.index in ignored_nodes:\n",
    "            continue\n",
    "\n",
    "        node_leaves = get_leaf_names(node)\n",
    "\n",
    "        if all_leaves.isdisjoint(node_leaves):\n",
    "            ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "        #\n",
    "        # if there aren't leaves from other phyla within this node it is monophyletic\n",
    "        if phylum1_leaves.intersection(node_leaves) and phylum1_leaves.issuperset(node_leaves):\n",
    "            phylum1_clades.add(node.index)\n",
    "            ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "        elif phylum2_leaves.intersection(node_leaves) and phylum2_leaves.issuperset(node_leaves):\n",
    "            phylum2_clades.add(node.index)\n",
    "            ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "    #\n",
    "    # now we add some flexibility to the monophyly of nodes for three reasons:\n",
    "    #   1) donor nodes within gene trees will never be monophyletic since the recipient\n",
    "    #      must be nested within it.\n",
    "    #   2) if there are other transfers from the donor and/or recipient phyla to a 3rd one\n",
    "    #      we still want to capture it.\n",
    "    #   3) good ol' phylogenetic uncertainty\n",
    "    phylum1_clades = merge_polyphyletic_clades(phylum1_clades, udag)\n",
    "    phylum2_clades = merge_polyphyletic_clades(phylum2_clades, udag)\n",
    "\n",
    "    #\n",
    "    # flag if we can identify relations between phyla:\n",
    "    #   1 nested within 2\n",
    "    #   2 nested within 1\n",
    "    #   1 and 2 are sisters\n",
    "    #\n",
    "    # if no relation is identified, leave as false\n",
    "    phyla_relationship_flag = 0\n",
    "\n",
    "    for clade1, clade2 in itertools.product(phylum1_clades, phylum2_clades):\n",
    "\n",
    "        clade1_ancestors = udag.vs[clade1].get_shortest_paths(udag.vs[0])[0][1:]\n",
    "        if clade2 in clade1_ancestors:\n",
    "            recipient_taxon         = fine_tune_recipient(dag.vs[clade1], \n",
    "                                                          related_phyla[0])\n",
    "            donor_taxon             = fine_tune_recipient(dag.vs[clade2], \n",
    "                                                          related_phyla[1])\n",
    "            phyla_relationship_flag = 1\n",
    "            \n",
    "            if (recipient_taxon, donor_taxon) != related_phyla:\n",
    "                descriptions[f'from {related_phyla[1]} to {related_phyla[0]}'].add('%s from %i to %i' % \\\n",
    "                                                                                   (candidate, donor_taxon, \n",
    "                                                                                    recipient_taxon))\n",
    "            else:\n",
    "                descriptions[f'from {related_phyla[1]} to {related_phyla[0]}'].add(candidate)\n",
    "            continue\n",
    "\n",
    "        clade2_ancestors = udag.vs[clade2].get_shortest_paths(udag.vs[0])[0][1:]\n",
    "        if clade1 in clade2_ancestors:\n",
    "            recipient_taxon         = fine_tune_recipient(dag.vs[clade2], \n",
    "                                                          related_phyla[1])\n",
    "            donor_taxon             = fine_tune_recipient(dag.vs[clade1], \n",
    "                                                          related_phyla[0])\n",
    "            phyla_relationship_flag = 1\n",
    "            \n",
    "            if (donor_taxon, recipient_taxon) !=  related_phyla:\n",
    "                descriptions[f'from {related_phyla[0]} to {related_phyla[1]}'].add('%s from %i to %i' % \\\n",
    "                                                                                   (candidate, donor_taxon, \n",
    "                                                                                    recipient_taxon))\n",
    "            else:\n",
    "                descriptions[f'from {related_phyla[0]} to {related_phyla[1]}'].add(candidate)\n",
    "            continue\n",
    "\n",
    "descriptions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fine_tune_recipient(dag.vs[clade1], related_phyla[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "candidate, donor_taxon, recipient_taxon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clade1, clade2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fine_tune_recipient(node, phylum):\n",
    "    node_taxids   = [int(leaf.split('.')[0]) for leaf in get_leaf_names(node)]\n",
    "    node_taxonomy = eggNOG_taxonomy.loc[node_taxids, \n",
    "                                             ['phylum', 'class', 'order', \n",
    "                                              'family', 'genus', 'species']\n",
    "                                            ].query(f'phylum=={phylum}')\n",
    "\n",
    "    most_specific_taxon = None\n",
    "    for rank, column in node_taxonomy.iteritems():\n",
    "        unique_taxa = column.unique()\n",
    "\n",
    "        if pd.notna(unique_taxa).all() and len(unique_taxa) == 1:\n",
    "            most_specific_taxon = unique_taxa[0]\n",
    "        else:\n",
    "            break\n",
    "    \n",
    "    return(most_specific_taxon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dag.vs[clade1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_taxids   = [int(leaf.split('.')[0]) for leaf in get_leaf_names(dag.vs[clade1])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_taxids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggNOG_taxonomy.loc[node_taxids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fine_tune_recipient(dag.vs[clade1], 200795)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "related_phyla = (200795, 1117)\n",
    "group_id = 'COG3259'\n",
    "cluster_num = '2'\n",
    "\n",
    "with cd('candidates/iqtree/'):\n",
    "    tmp_tree   = match_rooting(ete3.Tree('%s-cluster%s.treefile.rooted' % (group_id, \n",
    "                                                                           cluster_num),\n",
    "                                         format=1),\n",
    "                               ete3.Tree('%s-cluster%s.treefile'        % (group_id, \n",
    "                                                                           cluster_num),\n",
    "                                         format=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxids = {}\n",
    "for leaf in tmp_tree.get_leaf_names():\n",
    "    #\n",
    "    # leaf names are composed by <taxid>.<locus_tag>\n",
    "    #   ps: locus_tag may also have \".\" within it\n",
    "    tmp_taxid = int(leaf.split('.')[0])\n",
    "\n",
    "    if tmp_taxid not in taxids:\n",
    "        taxids[tmp_taxid] = []\n",
    "    taxids[tmp_taxid].append(leaf)\n",
    "\n",
    "#\n",
    "# select taxIDs from each assessed phylum...\n",
    "phylum1_taxonomy = eggNOG_taxonomy.loc[taxids].query('phylum==@related_phyla[0]')\n",
    "phylum2_taxonomy = eggNOG_taxonomy.loc[taxids].query('phylum==@related_phyla[1]')\n",
    "\n",
    "#\n",
    "# ... and their respect leaves\n",
    "phylum1_leaves = set()\n",
    "for taxid in phylum1_taxonomy.index:\n",
    "    phylum1_leaves.update(taxids[taxid])\n",
    "\n",
    "phylum2_leaves = set()\n",
    "for taxid in phylum2_taxonomy.index:\n",
    "    phylum2_leaves.update(taxids[taxid])\n",
    "\n",
    "all_leaves = phylum1_leaves.union(phylum2_leaves)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "dag  = tree_to_dag_bb_and_alrt(tmp_tree)          # we need a directed version...\n",
    "udag = dag.as_undirected(mode='each') # ... and an undirected one for different processes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylum1_clades = set() \n",
    "phylum2_clades = set()\n",
    "\n",
    "ignored_nodes  = [] # descendants of monophyletic nodes should be ignored once their\n",
    "                    #   ancestors have been stored.\n",
    "\n",
    "#\n",
    "# traverse through internal nodes\n",
    "for node in dag.vs:\n",
    "    if node.index in ignored_nodes:\n",
    "        continue\n",
    "\n",
    "    node_leaves = get_leaf_names(node)\n",
    "\n",
    "    if all_leaves.isdisjoint(node_leaves):\n",
    "        ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "    #\n",
    "    # if there aren't leaves from other phyla within this node it is monophyletic\n",
    "    if phylum1_leaves.intersection(node_leaves) and phylum1_leaves.issuperset(node_leaves):\n",
    "        phylum1_clades.add(node.index)\n",
    "        ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "    elif phylum2_leaves.intersection(node_leaves) and phylum2_leaves.issuperset(node_leaves):\n",
    "        phylum2_clades.add(node.index)\n",
    "        ignored_nodes.extend(get_descendant_indices(node, leaves=True))\n",
    "\n",
    "#\n",
    "# now we add some flexibility to the monophyly of nodes for three reasons:\n",
    "#   1) donor nodes within gene trees will never be monophyletic since the recipient\n",
    "#      must be nested within it.\n",
    "#   2) if there are other transfers from the donor and/or recipient phyla to a 3rd one\n",
    "#      we still want to capture it.\n",
    "#   3) good ol' phylogenetic uncertainty\n",
    "phylum1_clades = merge_polyphyletic_clades(phylum1_clades, udag)\n",
    "phylum2_clades = merge_polyphyletic_clades(phylum2_clades, udag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{36, 43, 48, 147}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phylum1_clades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dag.get_shortest_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[3]]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "udag.vs[72].shortest_paths(udag.vs[105])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
